src/gluonts/dataset/loader_v2.py

```diff
@@ -0,0 +1,189 @@
+
+    def __iter__(self):
+        for x in self.base_iterable:
+            if len(self.shuffle_buffer) < self.buffer_length:
+                self.shuffle_buffer.append(x)
+            if len(self.shuffle_buffer) == self.buffer_length:
+                yield self.sample_from_buffer()
+        while len(self.shuffle_buffer):
+            yield self.sample_from_buffer()
+
+
+class MultiProcessIterator(Iterator):
+    def __init__(
+        self,
+        base_iterable: Iterable,
+        num_workers: int,
+        num_entries: int,
+        max_queue_size: Optional[int] = None,
+    ):
+        assert num_workers >= 1
+        assert num_entries >= 0
+        assert max_queue_size is None or max_queue_size >= num_workers
+
+        self.base_iterable = base_iterable
+        self.num_workers = num_workers
+        self.num_entries = num_entries
+        self.max_queue_size = (
+            max_queue_size if max_queue_size is not None else 5 * num_workers
+        )
+
+        self.data_queue: Queue = Queue(maxsize=self.max_queue_size)
+        self.done_event = Event()
+        self.processes = []
+
+        for wid in range(self.num_workers):
+            p = Process(
+                target=self.worker_fn,
+                args=(
+                    wid,
+                    self.num_workers,
+                    self.base_iterable,
+                    self.data_queue,
+                    self.done_event,
+                ),
+            )
+            p.start()
+            self.processes.append(p)
+
+        self.count = 0
+
+    @staticmethod
+    def worker_fn(
+        worker_id: int,
+        num_workers: int,
+        iterable: Iterable,
+        data_queue: Queue,
+        end_event,
+    ):
+        MPWorkerInfo.worker_process = True
+        MPWorkerInfo.worker_id = worker_id
+        MPWorkerInfo.num_workers = num_workers
+
+        for entry in iterable:
+            if end_event.is_set():
+                break
+            data_queue.put((worker_id, entry))
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            wid, entry = self.data_queue.get(timeout=1.0)
+        except Empty:
+            raise StopIteration()
+
+        self.count += 1
+
+        if self.count == self.num_entries:
+            self.done_event.set()
+        elif self.count > self.num_entries:
+            raise StopIteration()
+
+        return entry
+
+    def _empty_queue(self):
+        try:
+            item = self.data_queue.get(block=False)
+            while item:
+                self.data_queue.get(block=False)
+        except Empty:
+            pass
+
+    def __del__(self):
+        # Send termination message to workers
+        self.done_event.set()
+        # Empty queue to make sure workers get the message
+        self._empty_queue()
+        for p in self.processes:
+            p.join()
+        self.data_queue.cancel_join_thread()
+        self.data_queue.close()
```

Comment:
So this method is not guaranteed to be called, just want to make sure you have taken this into consideration.

Reply:
Yes, good point. Things should now "gracefully" terminate without the need for this.